### ISSUE TITLE
Allow exporting the raw rules

### ISSUE THREAD

**Issue by 1player** [Smerity/texting_robots#1]
I've been considering using this library for my application as it saves me the effort of writing yet another robots.txt parser, but the only way of interfacing with the rule-set is through the `allowed` function. My app has its own custom URL matcher, so I would like to be able to export all the raw rules as exposed by robots (that match my user agent), and import them in my pattern matching logic. 

I'm thinking of just exposing a `rules()` function that returns the contents of the private `rules` field, possibly as an iterator of tuple `(Allow|Disallow, raw rule glob string)`. Would you be open to a PR?

**Comment by Smerity** [comment]
Thanks for the pull request - even including a test! =] I think this is a good idea but want to double check assumptions.

As you note, the raw `robots.txt` rule is likely what's most desired by library users.

The pull request is looking good but currently the `as_str` function can produce a different rule to the raw one as it might output the regex reduction (i.e. `/x***y/` is reduced to `/x*y/` as seen in [`test_robot_many_star_rule_simplifier`](https://github.com/Smerity/texting_robots/blob/f7a451a399789191f0623dd33b4f753f8e64acfd/src/test.rs#L323)).
By making `pattern` public in `MinRegex` (i.e. add `pub` [here](https://github.com/Smerity/texting_robots/blob/f7a451a399789191f0623dd33b4f753f8e64acfd/src/minregex.rs#L9)) we can expose the raw rule and `Robots::rules` could iterate over the `pattern` of each rule.

I presume exposing the raw rule, as seen in `robots.txt`, would be the most common use case.
Exposing the regex rule can be left until requested / later as it's also a tad more complicated given the fast regex path in `MinRegex`.

If you'd like to make the suggested change above, potentially adding the `/x***y/` case to your test to ensure we get back the original rule, that'd be great! Otherwise I'll get around to it when next I'm able.

Thanks for the contribution so far as this is definitely a useful addition.

**Comment by 1player** [comment]
Sure, I'll try and update my PR with your suggestion

**Comment by 1player** [comment]
But to be honest, I wouldn't mind, and perhaps prefer, to have the simplified pattern returned than the original one, if the `robots.txt` patterns are nonsense. 

I wonder if I should document that the `rules` function does actually return simplified patterns, and not the original ones verbatim? I'm not sure which approach is preferable.

**Comment by Smerity** [comment]
I may forge ahead with providing a way of accessing the raw rules as extracted from the `robots.txt` file whilst we think about the idea of returning a regex compatible rule.

The reason the regex compatible rule is complicated is, revisiting the code, there are actually three paths for rules:

- "startswith" matching (i.e. not a regex, just string matching)
- A star fast path which I've implemented to handle "/a/*/b" (and the pathological "/a/*****/b")
- A full regular expression when "$" is used

For your use case how would you want to use the boiled down rule? I could provide a way to convert the raw rules to reduced regular expressions in the rule itself. This is not yet done currently as those regular expressions are only constructed if they're necessary (i.e. they miss the other shortcut paths).

**Comment by 1player** [comment]
I honestly don't want to make this too complicated just for my use case, unless it's something you would like to add and expand yourself. My pattern matching code uses a slightly different syntax than robots, so the glob/regex hybrid the `rules()` function would spit out is good enough as I would have to convert it anyway. If your code already does some simplification (such as collapsing adjacent globs into one), that's even better. 

So in summary, for my use case, just having access to that `rules` function, which might not return what's present in robots.txt verbatim, is perfectly fine.

**Comment by Smerity** [comment]
I will close this for now only as it's unlikely to be a commonly called for need and requires rethinking how the internal ruleset is handled. I am very willing to reopen it in the future if either an elegant solution is found or if there's more call for it.

If a `rules` function that just returns the `robots.txt` rule would be optimal I can do that however.
